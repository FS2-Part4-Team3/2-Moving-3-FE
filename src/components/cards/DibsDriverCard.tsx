'use client';

import Image from 'next/image';
import { useEffect, useState } from 'react';
import standardProfile from '@/../public/assets/common/gnb/standard_profile.svg';
import like from '@/../public/assets/driver/ic_like.svg';
import star from '@/../public/assets/driver/ic_star_yellow.svg';
import { getDibsDriverListData } from '@/api/DriverService';
import { DibsDriverCardProps } from '@/interfaces/Card/DibsDriverCardInterface';
import SmallMovingTypeChips from '../chips/SmallMovingTypeChips';

export default function DibsDriverCard() {
  const [dibsDrivers, setDibsDrivers] = useState<DibsDriverCardProps | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const data = await getDibsDriverListData(1, 3);
        setDibsDrivers(data);
      } catch (error) {
        setError('데이터를 가져오는데에 실패했습니다.');
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, []);

  if (loading) {
    return <div>로딩중 ...</div>;
  }

  if (error) {
    return <div>{error}</div>;
  }

  if (!dibsDrivers?.totalCount) {
    return (
      <div className="w-[32.7rem] flex items-center justify-center pt-[3rem] font-medium text-[1.8rem] leading-[2.6rem] text-black-400">
        찜한 기사님이 없습니다.
      </div>
    );
  }

  return (
    <>
      {dibsDrivers.list.map((driver: any) => (
        <div className="w-[32.7rem] flex flex-col rounded-[1.6rem] border border-line-100 py-[1.6rem] px-[1.4rem] gap-[1.4rem] shadow-[-0.2rem_-0.2rem_1rem_rgba(220,220,220,0.3)]">
          <div className="flex gap-[0.4rem]">
            {driver.serviceType.map((item: string, index: number) => (
              <SmallMovingTypeChips key={index} type={item as 'SMALL' | 'HOME' | 'OFFICE' | 'APPOINTMENT' | 'WAITING'} />
            ))}
          </div>
          <p className="font-semibold text-[1.4rem] leading-[2.4rem] text-black-300">{driver.introduce}</p>
          <div className="w-full flex rounded-[0.6rem] border border-line-100 p-[1rem] gap-[1.2rem]">
            <div className="w-[4.6rem] h-[4.6rem]">
              {driver.image ? (
                <Image
                  src={driver.image}
                  alt="profile"
                  width={46}
                  height={46}
                  className="border-2 border-blue-400 rounded-full"
                />
              ) : (
                <Image
                  src={standardProfile}
                  alt="profile"
                  width={46}
                  height={46}
                  className="border-2 border-blue-400 rounded-full"
                />
              )}
            </div>
            <div className="w-full flex flex-col gap-[1.2rem]">
              <div className="flex items-center justify-between">
                <p className="font-semibold text-[1.4rem] leading-[2.4rem] text-black-300">{driver.name} 기사님</p>
                <div className="flex items-center gap-[0.2rem]">
                  <Image src={like} alt="like" width={24} height={24} />
                  <p className="font-medium text-[1.3rem] leading-[2.2rem] text-blue-400">{driver.favoriteCount}</p>
                </div>
              </div>
              <div className="flex items-center justify-between">
                <div className="flex gap-[0.2rem] items-center">
                  <Image src={star} alt="star" width={20} height={20} />
                  <p className="font-medium text-[1.3rem] leading-[2.2rem] text-black-300">{driver.score}</p>
                  <p className="font-medium text-[1.3rem] leading-[2.2rem] text-gray-300">({driver.reviewCount})</p>
                </div>
                <div className="h-[1.4rem] border border-line-200" />
                <div className="flex gap-[0.4rem]">
                  <p className="font-medium text-[1.3rem] leading-[2.2rem] text-gray-300">경력</p>
                  <p className="font-medium text-[1.3rem] leading-[2.2rem] text-black-300">{driver.career}년</p>
                </div>
                <div className="h-[1.4rem] border border-line-200" />
                <div className="flex gap-[0.4rem]">
                  <p className="font-medium text-[1.3rem] leading-[2.2rem] text-black-300">{driver.applyCount}건</p>
                  <p className="font-medium text-[1.3rem] leading-[2.2rem] text-gray-300">확정</p>
                </div>
              </div>
            </div>
          </div>
        </div>
      ))}
    </>
  );
}
