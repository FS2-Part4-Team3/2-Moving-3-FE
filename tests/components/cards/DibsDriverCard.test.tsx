import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import DibsDriverCard from '@/components/cards/DibsDriverCard';

jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    return <img {...props} />;
  },
}));

jest.mock('@/components/chips/SmallMovingTypeChips', () => {
  return function MockSmallMovingTypeChips({ type }: { type: string }) {
    return <div data-testid="moving-type-chip">{type}</div>;
  };
});

describe('DibsDriverCard', () => {
  const mockData = {
    id: 'driver-1',
    serviceType: ['HOME', 'SMALL'] as ('SMALL' | 'HOME' | 'OFFICE' | 'APPOINTMENT' | 'WAITING')[],
    introduce: '안녕하세요, 저는 경험이 많은 드라이버입니다.',
    image: 'http://placehold.it/56x56',
    name: '홍길동',
    favoriteCount: 10,
    score: 4.5,
    reviewCount: 20,
    career: 5,
    applyCount: 15,
  };

  it('render service type chip correctly', () => {
    render(<DibsDriverCard data={mockData} />);

    const chips = screen.getAllByTestId('moving-type-chip');
    expect(chips).toHaveLength(mockData.serviceType.length);
    mockData.serviceType.forEach((type, index) => {
      expect(chips[index]).toHaveTextContent(type);
    });
  });

  it('render driver introduction text', () => {
    render(<DibsDriverCard data={mockData} />);

    const intro = screen.getByText(mockData.introduce);
    expect(intro).toBeInTheDocument();
  });

  it('render driver profile information correctly', () => {
    render(<DibsDriverCard data={mockData} />);

    expect(screen.getByText(`${mockData.name} 기사님`)).toBeInTheDocument();
    expect(screen.getByText(mockData.favoriteCount)).toBeInTheDocument();
    expect(screen.getByText(mockData.score)).toBeInTheDocument();
    expect(screen.getByText(mockData.reviewCount)).toBeInTheDocument();
    expect(screen.getByText(`${mockData.career}년`)).toBeInTheDocument();
    expect(screen.getByText(`${mockData.applyCount}건`)).toBeInTheDocument();
  });

  it('render all images with correct attributes', () => {
    render(<DibsDriverCard data={mockData} />);

    const profileImage = screen.getByAltText('profile');
    expect(profileImage).toHaveAttribute('src', mockData.image);

    const likeIcon = screen.getByAltText('like');
    expect(likeIcon).toBeInTheDocument();

    const startIcon = screen.getByAltText('star');
    expect(startIcon).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<DibsDriverCard data={mockData} />);
    expect(container).toMatchSnapshot();
  });
});
