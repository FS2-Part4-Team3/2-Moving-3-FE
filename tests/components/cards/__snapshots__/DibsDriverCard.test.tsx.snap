// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DibsDriverCard matches snapshot 1`] = `
<div>
  <div
    class="w-[32.7rem] flex flex-col rounded-[1.6rem] border border-line-100 py-[1.6rem] px-[1.4rem] gap-[1.4rem] shadow-[-0.2rem_-0.2rem_1rem_rgba(220,220,220,0.3)]"
  >
    <div
      class="flex gap-[0.4rem]"
    >
      <div
        data-testid="moving-type-chip"
      >
        HOME
      </div>
      <div
        data-testid="moving-type-chip"
      >
        SMALL
      </div>
    </div>
    <p
      class="font-semibold text-[1.4rem] leading-[2.4rem] text-black-300"
    >
      안녕하세요, 저는 경험이 많은 드라이버입니다.
    </p>
    <div
      class="w-full flex rounded-[0.6rem] border border-line-100 p-[1rem] gap-[1.2rem]"
    >
      <div
        class="w-[4.6rem] h-[4.6rem]"
      >
        <img
          alt="profile"
          class="border-2 border-blue-400 rounded-full"
          height="46"
          src="http://placehold.it/56x56"
          width="46"
        />
      </div>
      <div
        class="w-full flex flex-col gap-[1.2rem]"
      >
        <div
          class="flex items-center justify-between"
        >
          <p
            class="font-semibold text-[1.4rem] leading-[2.4rem] text-black-300"
          >
            홍길동
             기사님
          </p>
          <div
            class="flex items-center gap-[0.2rem]"
          >
            <img
              alt="like"
              height="24"
              src="[object Object]"
              width="24"
            />
            <p
              class="font-medium text-[1.3rem] leading-[2.2rem] text-blue-400"
            >
              10
            </p>
          </div>
        </div>
        <div
          class="flex items-center justify-between"
        >
          <div
            class="flex gap-[0.2rem] items-center"
          >
            <img
              alt="star"
              height="20"
              src="[object Object]"
              width="20"
            />
            <p
              class="font-medium text-[1.3rem] leading-[2.2rem] text-black-300"
            >
              4.5
            </p>
            <p
              class="font-medium text-[1.3rem] leading-[2.2rem] text-gray-300"
            >
              (
              20
              )
            </p>
          </div>
          <div
            class="h-[1.4rem] border border-line-200"
          />
          <div
            class="flex gap-[0.4rem]"
          >
            <p
              class="font-medium text-[1.3rem] leading-[2.2rem] text-gray-300"
            >
              경력
            </p>
            <p
              class="font-medium text-[1.3rem] leading-[2.2rem] text-black-300"
            >
              5
              년
            </p>
          </div>
          <div
            class="h-[1.4rem] border border-line-200"
          />
          <div
            class="flex gap-[0.4rem]"
          >
            <p
              class="font-medium text-[1.3rem] leading-[2.2rem] text-black-300"
            >
              15
              건
            </p>
            <p
              class="font-medium text-[1.3rem] leading-[2.2rem] text-gray-300"
            >
              확정
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
